### PR TITLE
Add a more concise way to call the API functions

### PR DESCRIPTION
**Pull request by KristofferC**
When using the `Pkg.API` calls and wanting to e.g. add something by URL, it is a bit verbose to have to write

```jl
Pkg.add(PackageSpec(url = "..."))
```

Now it is possible to write

```jl
Pkg.add(url = "...")
```

in other words, relevant keyword arguments gets passed directly to the PackageSpec constructor. For multiple packages we can use a Vector of NamedTuples where each named tuple gets passed:

```jl
Pkg.add([(url = "...",), (url = "...",)])
```

Previously, all keyword arguments were passed directly to the `Context!` modifier but that is a real waste because no one uses that.

I need to expand the docs a bit and ideally this should be added for the `Pkg.Registry` module as well but that might be a later project...

### CODE REVIEW

**Comment by DilumAluthge** [comment]
Awesome!

File: src/Pkg.jl [hunk 1]
```diff
@@ -403,16 +404,26 @@ in the vector.
...
+
+!!! compat "Julia 1.4"
+    In Julia 1.4, many functions that take a `PackageSpec` or a `Vector{PackageSpec}` can be called with a more concise notation.
+    For example, `Pkg.add` can be called either as the explicit or concise versions as:
+
+    | Explicit                                                            | Concise
+    |:--------------------------------------------------------------------|:-----------------------------------------------|
+    | `Pkg.add(PackageSpec(name="Package))`                               | `Pkg.add(name = "Package")`                    |
+    | `Pkg.add(PackageSpec(url="www.myhost.com/MyPkg")))`                 | `Pkg.add(name = "Package")`                    |
+    |` Pkg.add([PacakgeSpec(name="Package"), PackageSpec(path="/MyPkg"])` | `Pkg.add([(name="Package",), (path="MyPkg",)])`|
```
Comment (DilumAluthge) [inline src/Pkg.jl]: Minor typo in the docstring: `PacakgeSpec`

**Comment by 00vareladavid** [comment]
I would strongly prefer to keep the API as dumb as possible. I believe we have generally agreed that if users want something more user friendly, they should just use the REPL.

**Comment by DilumAluthge** [comment]
I'd be on board with just using the REPL if pasting in REPL commands worked correctly.

But, for example, if you copy the following code and paste it into a Julia REPL:
```julia
] status
```

You'll get the error `ERROR: syntax: unexpected "]"`.

Given that copy-pasting into the Pkg REPL mode doesn't work, I feel like it'd be good to have a concise syntax for certain API functions so that I can tell a user to copy and paste this:
```julia
Pkg.add(url = "...")
```

And have it "just work".

**Comment by 00vareladavid** [comment]
I think copy and pasting is a separate issue. In any case, pkg"add ..." works just fine

**Comment by KristofferC** [comment]
> I would strongly prefer to keep the API as dumb as possible. 

Well, I would say that this is "dumber" than before.

>  I believe we have generally agreed that if users want something more user friendly, they should just use the REPL.

Not the way I see it, we should always try to be user-friendly. What we have agreed on is that it is ok for the REPL mode to be slightly ambiguous while the API should be very unambiguous. Nothing here changes the ambiguity level, it gives an arguably nicer way to write things.

**Comment by 00vareladavid** [comment]
> What we have agreed on is that it is ok for the REPL mode to be slightly ambiguous while the API should be very unambiguous.

Yes, that's a good distinction. Apologies, I didn't mean to put words in your mouth.

I'm just a bit wary of the error messages e.g. `Pkg.add(URL="foo")` does not hint that the kwarg is supposed to be `url`(lowercased) but `Pkg.add(Pkg.PackageSpec(URL="foo"))` does. I suppose if people want better error messages we can just tell them to use the more verbose option.

**Comment by 00vareladavid** [comment]
Closes #515?

**Comment by DilumAluthge** [comment]
Bump?

**Comment by DilumAluthge** [comment]
Bump. Whatever happened to this?

**Comment by KristofferC** [comment]
Rebased

**Pull request merged**